Prep for events

Berkyle opened the pull request:
# Description of Changes
<!-- Summarize your code change. -->

- Add events NavList items
- Use `Routes` component to separate routes 
   - e.g. Home and Events routes, eventually Administration and JBoard routes
- Organize style files into `_jss/` directories
- Wrapped application within the Header component (renamed to Navigation)
- Consolidate how contexts are organized
   - Consolidated the Events and User contexts with some patterns you followed for how context providers are exported
   - Moved Snackbar / Modal contexts into the same directory as the Events and User contexts
   - Stripped out unused contexts, including the AlertDialogueProvider and LoadingOverlayProvider (it's all source controlled so we can just add them back when they're needed)
   - Split Modal component and view logic into a component in src/components/common/Modal.tsx
- Consolidated some types

# Motivation & Context
<!-- Why this code change was made. -->
General preparation for more splitting up work between administration and events work. I also tore out some unused stuff. This is in part because a some of it was unused, convoluted, or just a bit unorganized, and in another part just to get more familiar with some of the code you had written.

# Type of change
<!-- Mark all that apply (the fewer the better) -->
- [ ] 🐛 Bug fix
- [x] ✨ New feature/Enhancement - 25%
- [ ] 🛠 Technical currency
- [x] 💎 Code cleanliness - 75%

# Snapshots
<!-- If not applicable, remove this section. -->
#### Nothing major, just added some space between the nav drawer and the content after nesting content within the navigation components. Also added the events nav items.
![prep_for_events](https://user-images.githubusercontent.com/34729254/85089371-2022e000-b1a8-11ea-80c9-aa52543e4140.gif)

# Discussion:
<!-- Any other context or topics to be brought to attention. -->

1. Should the navlinks automatically expand after being clicked?
2. I know moving the Modal component into the src/components/common/ directory may feel a little off, but I feel keeping everything with JSX logic and styles within the src/components/ directory is the right move.
   - That said, I'd like your eyes on the src/components/common/Modal.tsx component. I want to make sure I've re-implemented the closing transition logic as you'd like it.

juzerzarif committed:
✨ Changed drawer implementation to temporary mini variant

juzerzarif committed:
💎 Add tooling support for typescript paths and tree shaking

juzerzarif committed:
pr comments

juzerzarif committed:
Add linter and github ci

juzerzarif committed:
Fix linter warnings

juzerzarif committed:
Change buils workflow to report as status check

Berkyle committed:
💎 Organize jss & contexts. Prep for events.

Berkyle committed:
lint

Berkyle committed:
more lint

Berkyle commented:
the PR shows 9 commits because I branched off your PR (#3) before I started work

juzerzarif commented on src/components/common/Modal.tsx:
```diff
@@ -0,0 +1,78 @@
+
+  const handleClose = useCallback(() => dismiss(true), [dismiss]);
+
+  const isOpening = isOpen && !lastIsOpen;
+  const isClosing = !isOpen && lastIsOpen;
+
+  useEffect(() => {
+    if (isOpening) {
+      setIsVisible(true);
+      setTransitionIn(true);
+    } else if (isClosing) {
+      setTransitionIn(false);
+      setTimeout(() => setIsVisible(false), 400);
+    }
+  }, [isOpening, isClosing]);
```
You could just replace this with `isOpen`. It's a primitive value so `isOpen && lastIsOpen` and `!isOpen && !lastIsOpen` will never be true.

Berkyle replied:
couldn't, for example, `isOpen && lastIsOpen` be true if the modal is open and the `isLoading` prop changed?

juzerzarif replied:
Oh yeah I guess it will. My point was more you don't really need `lastOpen`, you could just straight up use `isOpen` in the effect

Berkyle replied:
addressing other comments later today but thanks for pointing that out, this helped me think through and clean up this component a lot [`a2149a1`](https://github.com/resident51/resident51/pull/4/commits/a2149a1fe2da5dc7372b9e8bb1fe2db9374aabc8)

juzerzarif commented on src/components/common/Modal.tsx:
```diff
@@ -0,0 +1,78 @@
+    }
+  }, [isOpening, isClosing]);
+
+  const content = isClosing ? lastChildren : children;
```
If I'm reading this correctly, this will keep the dismissed content rendered in the modal even after it's closed. When you pass in `isOpen = false` `isClosing` will be true and then it won't be false until `isOpen` is true again.

Berkyle replied:
I try to work around this by calling `setIsVisible(false)` within the useEffect above after the component reaches the `isClosing === true` state - do you think that is enough to get around this?

Berkyle replied:
In general though I know this component isn't as clear as it should be, definitely looking for ways to scoop out anything unnecessary.

juzerzarif replied:
Couldn't you just null out the content 400ms after you trigger the transition? You don't need the visible state, or really the transition in state for that matter. You just need a content state really.

Berkyle replied:
I did end up removing the transition in state, however the current implementation does depend on differentiating isOpen and isVisible (or something similar). This is to distinguish the "in" prop on the Material `<Grow in={...} ... />` component from the "open" prop on the `<MaterialModal open={...} ... />` component. Take a look at how our Modal component file is structured now, I think you'll find it slightly less bloated but I'm still looking for ways to improve it.

To your point about nulling out the content, this state is managed in the context right now and the component consumes it as the children prop and just holds on to the previous value, which I think may be the more lightweight and appropriate solution. What do you think?

juzerzarif replied:
The modal component already has a mechanic for waiting on a transition before closing with the `closeAfterTransition` props. All the hijinks with holding onto children for one extra render for a certain case is more convoluted than it needs to be. All you really need to do is accept content state from the parent while it's open and null it out once you're done with your transition. Here's what I mean: https://gist.github.com/juzerzarif/a9eaef306979ba4016b1168c83e1997d

Berkyle replied:
ahhh I should have just looked at the Grow API, that `onExited` props is exactly what I needed - nice find. Check out https://github.com/resident51/resident51/pull/4/commits/b6054496562b79fcb9c1e9a202ff4e7d7bf1fbde

A couple notes:
- Added the `disableRestoreFocus` prop because by default the modal returns focus to the triggering element, in this case leaving the ripple effect on the Login button
- could you use a branch instead of a gist next time? it makes viewing changes easier, plus I can checkout the branch to view the change and cherry-pick

juzerzarif commented on src/components/navigation/NavigationItems.tsx:
```diff
@@ -2,13 +2,31 @@ import React from 'react';
 
 import {
   BusinessCenter as BusinessCenterIcon,
+  CalendarTodayOutlined,
   Feedback as FeedbackIcon,
   GitHub as GithubIcon,
```
If we wanna have outlined icons, we should change them all to be outlined

Berkyle replied:
good point, replaced in c1be1406441c7d554e05b14f660e9d69c8665206

juzerzarif commented on src/components/navigation/AuthActionButton.tsx:
```diff
@@ -5,29 +5,25 @@ import Button from '@material-ui/core/Button';
-import { ModalCtx } from '@app/types';
-
-import { useModal } from '@app/contexts/ui/ModalProvider';
+import { useModal } from '@app/contexts/Modal';
```
I would keep this in the ui directory. It's a ui context even if it doesn't have any ui in it.

Berkyle replied:
I don't think we'll need to organize our contexts that much until we hit 10-12 distinct context providers, even if some are UI-specific and others aren't. Directory organization is purely subjective of course, but do you feel this should be done right away?

juzerzarif replied:
I think if you're willing to do something at 10-12 files you should just do it to begin with.

Berkyle replied:
I don't think it's necessary to preemptively organize components like that, and at this point the majority of our contexts would go in a `ui/` folder anyway so why not just keep them one directory up? This may just be where we differ in organization preferences but I'd prefer to stray away from nesting things that conceptually aren't that different.

Berkyle replied:
leaving this unresolved though because there certainly could be more discussion on this

juzerzarif replied:
My point was this: say we have 20 different contexts in that directory, would you still make the same argument of not separating them because they're all contexts at the end of the day? If not, then when you do make that change down the road it will bring along a refactor with it. So why not just start breaking them out now? I guess it's essentially the same argument of if you know you're going to have to do something down the road just do it now

Berkyle replied:
That same point could be made about every file in this project. Why not divvy up the top-level components in `src/components` further? Won't we eventually have like 20 top-level route components? Of course, but hypotheticals don't make sense here - we _don't_ have that many right now, and we _don't_ have 20, or 10, contexts, even if the old contexts were still there. Plus, there is no implicit refactor involved in waiting; Code will instantly and reliably update imports for us.

My answer to "why not just do this now" is the same reason I didn't think refactoring the other context utilities was profitable. If we try to address use cases that literally do not exist yet, we are guaranteed to introduce tech debt. The odds of creating the ideal solution for a problem we don't fully understand aren't just unlikely, it's utterly impossible. Waiting until the problem exists is both the most convenient and the most logical approach. Sure, that's pretty dramatic for organizing files, but the philosophy is the same. There isn't a reason to solve a problem that long before it exists.

Berkyle replied:
was going to there is no refactor involved because code can fix imports for us but this doesn't seem to be the case with import aliasing. Maybe there's a fix but for now, sure, check out https://github.com/resident51/resident51/pull/4/commits/736686c9facfd3dd9c3c1760e753283feb4a5df9

juzerzarif commented:
I don't think the nav links need to be expanded on open. It's an unnecessary animation that I don't want to do unless the user triggers it.

juzerzarif commented:
As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.

Berkyle committed:
replace outlined nav icon

Berkyle commented:
> As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.

I don't doubt we'll likely need a loading overlay, alert dialogue, etc. at some point, but I think it will be best to wait until we know how the entire app will be laid out and what the exact use case of those utilities will be before we make assertions about how those components behave. While I agree it's useful to think ahead, I feel keeping them now just adds upkeep that we don't need during this phase of development.

juzerzarif commented:
> > As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.
> 
> I don't doubt we'll likely need a loading overlay, alert dialogue, etc. at some point, but I think it will be best to wait until we know how the entire app will be laid out and what the exact use case of those utilities will be before we make assertions about how those components behave. While I agree it's useful to think ahead, I feel keeping them now just adds upkeep that we don't need during this phase of development.

I don't really see what "upkeep" there'd be for just keeping a file. You could change it if need be, but I just don't really see the point of removing those things. If you were making a case that it didn't need to be implemented right now, sure maybe. But it's already there and again if you're planning on just putting it back in, then why are you taking it out.

juzerzarif commented:
The modal content disappears before the modal closes. I'm assuming that happens because the isOpen prop is changed before the children are nulled.
![2020-06-21_15-44-11](https://user-images.githubusercontent.com/22772637/85234983-0a473200-b3d7-11ea-8c5a-93a9b083fb0b.gif)

juzerzarif commented:
The DOM layout is a little wrong: The main element is pushed all the way to the top and there are two toolbar offsets: one inside the main element and one outside (with no width). The margin left for the slide menu is slightly shorter than the slide menu. 

Need to:
1) Remove the toolbar offset inside the main element
2) Add width:100% to the offset outside
3) Add flex-flow:column to the root flexbox
4) Increase the margin left to have 7 spacing units

![image](https://user-images.githubusercontent.com/22772637/85235392-25677100-b3da-11ea-86e8-57ec55fb8749.png)

Berkyle committed:
fix modal dismiss

Berkyle committed:
fixup

Berkyle commented:
> The modal content disappears before the modal closes. I'm assuming that happens because the isOpen prop is changed before the children are nulled.

Cleaned up `src/components/common/Modal.tsx` and fixed this, needed some time away from it i guess. 

![fix-modal](https://user-images.githubusercontent.com/34729254/85432129-21f1f800-b548-11ea-93da-5071efc91498.gif)
_I'm not sure why giphy is showing that padlock icon as grey instead of pink._

Berkyle commented:
> > > As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.
> > 
> > 
> > I don't doubt we'll likely need a loading overlay, alert dialogue, etc. at some point, but I think it will be best to wait until we know how the entire app will be laid out and what the exact use case of those utilities will be before we make assertions about how those components behave. While I agree it's useful to think ahead, I feel keeping them now just adds upkeep that we don't need during this phase of development.
> 
> I don't really see what "upkeep" there'd be for just keeping a file. You could change it if need be, but I just don't really see the point of removing those things. If you were making a case that it didn't need to be implemented right now, sure maybe. But it's already there and again if you're planning on just putting it back in, then why are you taking it out.

It's less about upkeep and more that it is not being used. You're right that it costs nothing to just leave it there and I am sure it will be useful but while we're figuring out project structure I'd like a little less to move around and worry about. I know it probably took a while to create but right now it's just grinding against some OCD - that's not a technical reason of course so I definitely appreciate your flexibility haha.

Berkyle commented:
> The DOM layout is a little wrong: The main element is pushed all the way to the top and there are two toolbar offsets: one inside the main element and one outside (with no width). The margin left for the slide menu is slightly shorter than the slide menu.
>  [...]

Wow I really just didn't see we had two toolbar offsets, huh. Good eye, thank you

![Screen Shot 2020-06-23 at 12 31 06 PM](https://user-images.githubusercontent.com/34729254/85435663-6e8c0200-b54d-11ea-8889-b1967a272c39.png)

Berkyle committed:
remove duplicate toolbar offset

juzerzarif commented:
> > > > As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.
> > > 
> > > 
> > > I don't doubt we'll likely need a loading overlay, alert dialogue, etc. at some point, but I think it will be best to wait until we know how the entire app will be laid out and what the exact use case of those utilities will be before we make assertions about how those components behave. While I agree it's useful to think ahead, I feel keeping them now just adds upkeep that we don't need during this phase of development.
> > 
> > 
> > I don't really see what "upkeep" there'd be for just keeping a file. You could change it if need be, but I just don't really see the point of removing those things. If you were making a case that it didn't need to be implemented right now, sure maybe. But it's already there and again if you're planning on just putting it back in, then why are you taking it out.
> 
> It's less about upkeep and more that it is not being used. You're right that it costs nothing to just leave it there and I am sure it will be useful but while we're figuring out project structure I'd like a little less to move around and worry about. I know it probably took a while to create but right now it's just grinding against some OCD - that's not a technical reason of course so I definitely appreciate your flexibility haha.

Alright fuck it sure. If you're gonna do that what I would like to see in the commit tree is a single clean commit for each of the contexts removed that can be cherry-picked at a later date to add them in.

juzerzarif commented on src/contexts/Modal.tsx:
```diff
@@ -0,0 +1,73 @@
+    setIsOpen(true);
+  }, []);
+
+  const dismiss = useCallback((skipDismissNotification?: boolean) => {
```
I know I wrote this but I think it's better to just get rid of that parameter, I don't really see a logical case for it now and I think it just came along from another implementation we had for this at work

Berkyle replied:
donezo https://github.com/resident51/resident51/pull/4/commits/1810cf03762f138a810a74612cddc138cf602e38

juzerzarif commented on src/contexts/Modal.tsx:
```diff
@@ -0,0 +1,73 @@
+  }, []);
+
+  return (
+    <ModalContext.Provider value={{ isOpen, disclose, dismiss }}>
```
It might be worth memoizing the context object. I know that the individual fields are memoized but I wouldn't really want re-renders of this component outside of its scope triggering its consumers. This isn't that big a concern since this sits pretty high in the tree right now but if we end up using it nested somewhere, it could become an issue.

Berkyle replied:
sure, done here https://github.com/resident51/resident51/pull/4/commits/0e988a4d25c53ee7deaf79828486af9aaf40d7b9

Berkyle committed:
simplify modal component

Berkyle committed:
remove unused dismiss param

Berkyle committed:
add back contexts and types

Berkyle committed:
remove alert dialogue provider

Berkyle committed:
remove loading overlay provider

Berkyle committed:
memoize modal context value

Berkyle commented:
> > > > > As far as stripping out the rest of the ui contexts, I put those in as part of a basic framework that can be leveraged as we build ui moving forward. These are things that will definitely be needed in the future so I don't quite see the point of removing things that we know we're gonna have to put back in.
> > > > 
> > > > 
> > > > I don't doubt we'll likely need a loading overlay, alert dialogue, etc. at some point, but I think it will be best to wait until we know how the entire app will be laid out and what the exact use case of those utilities will be before we make assertions about how those components behave. While I agree it's useful to think ahead, I feel keeping them now just adds upkeep that we don't need during this phase of development.
> > > 
> > > 
> > > I don't really see what "upkeep" there'd be for just keeping a file. You could change it if need be, but I just don't really see the point of removing those things. If you were making a case that it didn't need to be implemented right now, sure maybe. But it's already there and again if you're planning on just putting it back in, then why are you taking it out.
> > 
> > 
> > It's less about upkeep and more that it is not being used. You're right that it costs nothing to just leave it there and I am sure it will be useful but while we're figuring out project structure I'd like a little less to move around and worry about. I know it probably took a while to create but right now it's just grinding against some OCD - that's not a technical reason of course so I definitely appreciate your flexibility haha.
> 
> Alright fuck it sure. If you're gonna do that what I would like to see in the commit tree is a single clean commit for each of the contexts removed that can be cherry-picked at a later date to add them in.

fair enough - check out https://github.com/resident51/resident51/pull/4/commits/b5808ad03c2f5704cb8bcd63d90d0e3e1d0a4787, https://github.com/resident51/resident51/pull/4/commits/3c99f91f420f62f30e000efbbb065b1f0c18dbc4, and https://github.com/resident51/resident51/pull/4/commits/130d368b9ae8aa61e5673c4f98f764a515de9626

Berkyle committed:
divide context directories

Berkyle committed:
move src/contexts/utils -> src/contexts/services/utils
